Move ESMF-based surface regridding code from GDASApp to UFS_UTILS

ClaraDraper-NOAA opened the pull request:
## DESCRIPTION OF CHANGES: 

Added a new directory, with the surface regridding code that was in [GDASApp](https://github.com/NOAA-EMC/DA-utils/tree/develop/src/regridStates). 

Changes to the files, from what was in GDASApp: 

-Added a regression test for the regridding code, which has been tested on hera only. The regression test inputs and baseline files are here: 

/scratch2/BMC/gsienkf/Clara.Draper/gerrit-hera/PRprep/new_test

-Added Doxygen documentation files, plus necessary changes to the fortran routines.

There are are no changes to the executable code in the fortran files.

## TESTS CONDUCTED: 
If there are changes to the build or source code, the tests below must be conducted. Contact a repository manager if you need assistance.

- [x] Compile branch on all Tier 1 machines using Intel (Orion, Jet, Hera, Hercules and WCOSS2). Done using 009456b.
- [x] Compile branch on Hera using GNU. Done using 009456b.
- [x] Compile branch in 'Debug' mode on WCOSS2. Done using 009456b.
- [x] Compile with Doxygen on any machine with no errors. For details, see DOCUMENTATION section.
- [x] Run unit tests locally on any Tier 1 machine.  Done on Hera using 9f313c7 and GNU. All tests passed.
- [x] Run new `regrid_sfc` consistency test locally on all Tier 1 machines. Done successfully using 009456b.

Describe any additional tests performed.

Also, checked that the regridding executable can be run from global_workflow. This will require a spack-stack [update](https://github.com/JCSDA/spack-stack/issues/1562#issuecomment-2725573898
) 
## DEPENDENCIES:
None. 

## DOCUMENTATION:
All new and updated source code must be documented with Doxygen.

- [x] Doxygen is updated.  Using 009456b, the Doxygen was checked and looked correct. See: [here](https://www.emc.ncep.noaa.gov/jcsda/ggayno/doxygen/html/regrid_sfc/index.html)

## ISSUE: 
 Fixes issue [1032](https://github.com/ufs-community/UFS_UTILS/issues/1032)

ClaraDraper-NOAA committed:
Moved regrid_sfc utility to UFS_UTILS, from GDASApp/DA-utils.
Compilation, installation, and links are working.

ClaraDraper-NOAA committed:
Switching F90 to f90.

ClaraDraper-NOAA committed:
Added freeform flag to gcc.

ClaraDraper-NOAA committed:
Fixed bug in resolution in script, updated compilation flags.

ClaraDraper-NOAA committed:
changes for intel compilation.

ClaraDraper-NOAA committed:
Cmake bug

ClaraDraper-NOAA committed:
added regression test

ClaraDraper-NOAA committed:
Tidied up the regression test.

ClaraDraper-NOAA committed:
Tidied up the regression test.

ClaraDraper-NOAA committed:
Merge branch 'feature/mv_regridding' of https://github.com/ClaraDraper-NOAA/UFS_UTILS into feature/mv_regridding

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/mv_regridding

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/mv_regridding

ClaraDraper-NOAA committed:
Added Doxygen

ClaraDraper-NOAA committed:
Updates to Doxygen

ClaraDraper-NOAA committed:
Added regression test drivers for the other machines.

ClaraDraper-NOAA commented:
@GeorgeGayno-NOAA Creating as a draft, as I haven't done the unit tests for the new executable. I'm hoping you can still help me with that. Everything else is ready to go (I hope!).

GeorgeGayno-NOAA commented on reg_tests/regrid_sfc/driver.s4.sh:
Do you work on S4?

ClaraDraper-NOAA replied:
no.

GeorgeGayno-NOAA replied:
s4 is not a Tier 1 platform, which means we don't have to support it unless there is a reason to. So, go ahead and remove the 's4' logic.

ClaraDraper-NOAA replied:
done.

GeorgeGayno-NOAA commented:
@ClaraDraper-NOAA - I know that other consistency tests have separate driver scripts for each machine, but one of my goals is to combine them if possible. Please see my script on Hera: `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara/reg_tests/regrid_sfc/driver.sh`. 

If that script works on the RDHPCS machines, I can expand it for WCOSS2.

ClaraDraper-NOAA committed:
Merge branch 'feature/mv_regridding' of https://github.com/ClaraDraper-NOAA/UFS_UTILS into feature/mv_regridding

ClaraDraper-NOAA committed:
George's unified driver for RDHPCS

ClaraDraper-NOAA commented:
@GeorgeGayno-NOAA The driver.sh script works for hera. I've added it, and deleted the separate scripts.

GeorgeGayno-NOAA commented:
@ClaraDraper-NOAA - I can't get the consistency test to run because the `/sfcincr_gsi` file is missing.  I can also set up the input and baseline directories on each machine.

ClaraDraper-NOAA committed:
Deleted s4.

ClaraDraper-NOAA commented:
> @ClaraDraper-NOAA - I can't get the consistency test to run because the `/sfcincr_gsi` file is missing. I can also set up the input and baseline directories on each machine.

The sfcinc file is here: 
/scratch2/BMC/gsienkf/Clara.Draper/gerrit-hera/PRprep/ufs_utils.fd/reg_tests/regrid_sfc/input_data_noahmp/

ClaraDraper-NOAA commented:
input only.

ClaraDraper-NOAA commented:
I see what the problem is. I'm fixing it (and also trying to figure out why my tests are passing, when they shouldn't have).

ClaraDraper-NOAA committed:
Fixed file name in consistent check for regrid_sfc.

ClaraDraper-NOAA commented:
I fixed both issues.

GeorgeGayno-NOAA commented:
Can I move the 'fix' files to the role directory?  For example, on Hera, see: `/scratch1/NCEPDEV/nems/role.ufsutils/ufs_utils/reg_tests/regrid_sfc/fix`

Since the official 'fix' files will change periodically, your test might start failing after a future upgrade.

I update the `driver.sh` and `gauss2fv3incr.sh` accordingly. Find them here: `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara/reg_tests/regrid_sfc`

ClaraDraper-NOAA commented:
Looks good to me, thanks @GeorgeGayno-NOAA.  ( I assume you know, but I checked the other tests and they're a mix of both options).

GeorgeGayno-NOAA commented:
I wrote a unit test for the namelist read. Find it on Hera: `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara`

- Baseline all new files under `./tests/regrid_sfc`
- Baseline the updated file: `./tests/CMakeLists.txt`
- I pulled the namelist read routine to its own file: `./sorc/regrid_sfc.fd/readin_setup.F90`
- The following files were modified: `./sorc/regrid_sfc.fd/regridStates.F90` and `CMakeLists.txt`

I am not sure how to unit test the rest of your code, as it is ESMF calls. Maybe the consistency test is sufficient.

ClaraDraper-NOAA commented:
> I wrote a unit test for the namelist read. Find it on Hera: `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara`
> 
> * Baseline all new files under `./tests/regrid_sfc`
> * Baseline the updated file: `./tests/CMakeLists.txt`
> * I pulled the namelist read routine to its own file: `./sorc/regrid_sfc.fd/readin_setup.F90`
> * The following files were modified: `./sorc/regrid_sfc.fd/regridStates.F90` and `CMakeLists.txt`
> 
> I am not sure how to unit test the rest of your code, as it is ESMF calls. Maybe the consistency test is sufficient.

Thanks @GeorgeGayno-NOAA. It was also the ESMF calls that I wasn't sure how to navigate. This is enough for me. Do you want me to add it in?

ClaraDraper-NOAA committed:
Added unit test from George.

ClaraDraper-NOAA commented:
> > > I wrote a unit test for the namelist read. Find it on Hera: `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara`
> > > 
> > > * Baseline all new files under `./tests/regrid_sfc`
> > > * Baseline the updated file: `./tests/CMakeLists.txt`
> > > * I pulled the namelist read routine to its own file: `./sorc/regrid_sfc.fd/readin_setup.F90`
> > > * The following files were modified: `./sorc/regrid_sfc.fd/regridStates.F90` and `CMakeLists.txt`
> > > 
> > > I am not sure how to unit test the rest of your code, as it is ESMF calls. Maybe the consistency test is sufficient.
> > 
> > 
> > Thanks @GeorgeGayno-NOAA. It was also the ESMF calls that I wasn't sure how to navigate. This is enough for me. Do you want me to add your test to my branch?
> 
> Yes.

I've added it and checked it passed on hera. Converting this to ready to review. Thanks @GeorgeGayno-NOAA !

GeorgeGayno-NOAA commented on sorc/regrid_sfc.fd/gauIncr2native.nml:
Is this namelist file needed in the source code directory?

ClaraDraper-NOAA replied:
With the new regression test, no. I removed it.

GeorgeGayno-NOAA commented on sorc/regrid_sfc.fd/README:
I would move this information to the ./docs/regrid_sfc.md file. That way it becomes part of the Doxygen. Then remove this README file.

ClaraDraper-NOAA replied:
done.

GeorgeGayno-NOAA commented:
When I move the namelist read routine to its own routine, I did not get the Doxygen correct. Please baseline my updated version: `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara/sorc/regrid_sfc.fd/readin_setup.F90`

ClaraDraper-NOAA committed:
Fixed Doxygen, deleted namelist.

ClaraDraper-NOAA commented:
> When I moved the namelist read routine to its own routine, I did not get the Doxygen correct. Please baseline my updated version: `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara/sorc/regrid_sfc.fd/readin_setup.F90`

 done.

ClaraDraper-NOAA committed:
Moved readme into doxygen.

ClaraDraper-NOAA commented:
@GeorgeGayno-NOAA thanks. I've addressed all your comments.

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA thanks. I've addressed all your comments.

The Doxygen looks good. See: https://www.emc.ncep.noaa.gov/jcsda/ggayno/doxygen/html/regrid_sfc/index.html

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/mv_regridding

GeorgeGayno-NOAA commented:
> > > @GeorgeGayno-NOAA thanks. I've addressed all your comments.
> > 
> > 
> > The Doxygen looks good. See: https://www.emc.ncep.noaa.gov/jcsda/ggayno/doxygen/html/regrid_sfc/index.html
> 
> Thanks George. What's the next step?

Please baseline my updates to the regression test in `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara`

- `./reg_tests/rt.sh`
- `./reg_tests/regrid_sfc/driver.sh`
- `./reg_tests/regrid_sfc/gauss2fv3incr.sh`

ClaraDraper-NOAA committed:
George's test updates.

ClaraDraper-NOAA commented:
> > > > @GeorgeGayno-NOAA thanks. I've addressed all your comments.
> > > 
> > > 
> > > The Doxygen looks good. See: https://www.emc.ncep.noaa.gov/jcsda/ggayno/doxygen/html/regrid_sfc/index.html
> > 
> > 
> > Thanks George. What's the next step?
> 
> Please baseline my updates to the regression test in `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara`
> 
> * `./reg_tests/rt.sh`
> * `./reg_tests/regrid_sfc/driver.sh`
> * `./reg_tests/regrid_sfc/gauss2fv3incr.sh`

Done.

GeorgeGayno-NOAA commented:
Will do some final tests, then merge.

GeorgeGayno-NOAA commented:
I tried running the consistency test on WCOSS2 in 'debug' mode and it stopped with an array mismatch:

```
forrtl: severe (408): fort: (33): Shape mismatch: The extent of dimension 2 of array LON_PTR_COORD is 65 and 
the corresponding extent of array LON_PTR_FIELD is 64

Image              PC                Routine            Line        Source
regridStates.x     00000000020CB74F  Unknown               Unknown  Unknown
regridStates.x     000000000043DC75  grids_io_mp_creat         647  grids_IO.F90
regridStates.x     00000000004294A1  grids_io_mp_setup          72  grids_IO.F90
regridStates.x     000000000042368B  MAIN__                    107  regridStates.F90
```

This is the line in the `grids_IO.F90`:

```
     ! set coord pointe to field pointer
     lon_ptr_coord = lon_ptr_field
```
It only happens on PET 5. On all other PETS, dimension 2 is 64 for both pointers. LAT_PTR_COORD has the same problem on PET 5.

GeorgeGayno-NOAA commented:
I forgot to update the consistency test driver script for WCOSS2. I placed the updated version on Hera: `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.clara/reg_tests/regrid_sfc/driver.sh`

ClaraDraper-NOAA committed:
George's updates to the reg_test on WCOSS.

GeorgeGayno-NOAA commented on sorc/regrid_sfc.fd/grids_IO.F90:
```diff
@@ -0,0 +1,681 @@
+ character(len=15), dimension(2) :: latvar_list
+
+ ! center coord names
+ lonvar_list(1) = 'grid_center_lon'
```
Here you are specifying center and corner variables.

GeorgeGayno-NOAA commented on sorc/regrid_sfc.fd/grids_IO.F90:
```diff
@@ -0,0 +1,681 @@
+ latvar_list(2) = 'grid_corner_lat'
+
+ ! Create the fields
+ do v=1,2
```
In this loop, you are only specifying 'center' fields.

GeorgeGayno-NOAA commented on sorc/regrid_sfc.fd/grids_IO.F90:
```diff
@@ -0,0 +1,681 @@
+ call lonlat_read_into_fields(localpet, i_dim, j_dim,  &
+                         trim(grid_setup%fname_coord), trim(grid_setup%dir_coord), &
+                         grid_setup, gauss_grid, lon_fields, lat_fields)
+
```
In this loop, you specify 'center' for v=1 and 'corner' for v=2. This is leading to the mismatch. All 'corner' arrays should contain one extra row with dimension 768x385. That is why I am seeing a j dimension of 65 on PET 5 and 64 on all other PETS.

ClaraDraper-NOAA replied:
Nice work @GeorgeGayno-NOAA ! I'll get it fixed.

ClaraDraper-NOAA committed:
Fix corner / center bug.

ClaraDraper-NOAA committed:
update grid_setup to directly use scrip file.
-results in 10**-15 sig changes in the test case.

ClaraDraper-NOAA committed:
Merge branch 'develop' into feature/mv_regridding

ClaraDraper-NOAA commented:
@GeorgeGayno-NOAA I've fixed the bug, and then wound up replacing that section of code with setting-up the gaussian grid directly from the scrip file, which is much neater. I don't have access to WCOSS, but this now compiles in debug mode on hera (previous didn't). It fails the reg_test, but the differences are very minor (< 10**-15).  I also checked that it's wrapping correctly on 0 longitude (i.e., correctly using a periodic dimenson).

ClaraDraper-NOAA commented:
New test output: 
/scratch2/BMC/gsienkf/Clara.Draper/gerrit-hera/PRprep/new_test/regrid_sfc/baseline_data

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA I've fixed the bug, and then wound up replacing that section of code with setting-up the gaussian grid directly from the scrip file, which is much neater. I don't have access to WCOSS, but this now compiles in debug mode on hera (previous didn't). It fails the reg_test, but the differences are very minor (< 10**-15). I also checked that it's wrapping correctly on 0 longitude (i.e., correctly using a periodic dimenson).

I was able to compile and run the tests on WCOSS2 in 'debug' mode. I noted the same minor differences from the baseline.

ClaraDraper-NOAA committed:
Enable gaussian grid setup for any multiple of 6 tasks.

ClaraDraper-NOAA committed:
Tidy up.

ClaraDraper-NOAA commented:
@GeorgeGayno-NOAA I think this is ready to go. I changed the pet distribution for the Gaussian case so that the npets does not need to be a factor of the longitudinal dimension (which was causing workflow problems). It does now need to be a multiple of 6 though - and the Gaussian longitudinal dimension also needs to be a multiple of 6 - this is true for all our grids, except C128, which has Gaussian output at 512x256. 
@aerorahul suggested we merge this now, then fix that later. 

Note: the results are not reproducible if we change the number of tasks. With 6 tasks at C192, the wall time is 6 seconds (5 seconds with 36 tasks). I suspect that this is not a new problem. For the workflow, I suggest we proceed for now, and fix the tasks at 6 for reproducibility. @CatherineThomas-NOAA @aerorahul  - does this sound OK? 

This passes the test, and compiles and runs in debug on hera. 

More info: 
The issue with the restrictions on the number of tasks is coming from the fact that I can't get the unabalanced DE distributions in the CreateGrid routines to work - and I see from the comments in the code @GeorgeGayno-NOAA  sent me that this was also a problem for the setting up the cube sphere grid (which is where the multiples of 6 requirements is coming from). I can probably fix this later, but if my proposal above is not OK and we need it fixed sooner, I'll need some help from a software specialist.

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA I think this is ready to go. I changed the pet distribution for the Gaussian case so that the npets does not need to be a factor of the longitudinal dimension (which was causing workflow problems). It does now need to be a multiple of 6 though - and the Gaussian longitudinal dimension also needs to be a multiple of 6 - this is true for all our grids, except C128, which has Gaussian output at 512x256. @aerorahul suggested we merge this now, then fix that later.
> 
> Note: the results are not reproducible if we change the number of tasks. With 6 tasks at C192, the wall time is 6 seconds (5 seconds with 36 tasks). I suspect that this is not a new problem. For the workflow, I suggest we proceed for now, and fix the tasks at 6 for reproducibility. @CatherineThomas-NOAA @aerorahul - does this sound OK?
> 
> This passes the test, and compiles and runs in debug on hera.
> 
> More info: The issue with the restrictions on the number of tasks is coming from the fact that I can't get the unabalanced DE distributions in the CreateGrid routines to work - and I see from the comments in the code @GeorgeGayno-NOAA sent me that this was also a problem for the setting up the cube sphere grid (which is where the multiples of 6 requirements is coming from). I can probably fix this later, but if my proposal above is not OK and we need it fixed sooner, I'll need some help from a software specialist.

In the interest of time, I will run the final tests, then merge what we have.
